File: app/models/device.py
"""
Device model.

Represents network devices discovered during scans.
"""

from typing import TYPE_CHECKING, Optional

from sqlmodel import Field, Relationship

from app.models.base import BaseModel

if TYPE_CHECKING:
    from app.models.category import Category
    from app.models.location import Location
    from app.models.mac import Mac
    from app.models.owner import Owner

class Device(BaseModel, table=True):
    """Device model representing a network device.""" 
    name: str | None = Field(default=None, unique=True, max_length=200)
    model: str | None = Field(default=None, max_length=200)

    category_id: int | None = Field(default=None, foreign_key="category.id")
    location_id: int | None = Field(default=None, foreign_key="location.id")
    owner_id: int | None = Field(default=None, foreign_key="owner.id")

    category: Optional["Category"] = Relationship(back_populates="devices")
    location: Optional["Location"] = Relationship(back_populates="devices")
    owner: Optional["Owner"] = Relationship(back_populates="devices")
    macs: list["Mac"] = Relationship(back_populates="device")

    @property
    def primary_mac(self) -> Optional["Mac"]:
        """Get the primary MAC address (most recently seen)."""
        if not self.macs:
            return None
        return max(self.macs, key=lambda mac: mac.last_seen or 0)

File: frontend/src/app/interfaces/device.ts
import { Category } from '#interfaces/category';
import { Location } from '#interfaces/location';
import { Mac } from '#interfaces/mac';
import { Owner } from '#interfaces/owner';

export interface Device {
  id: number;
  name: string | null;
  model: string | null;
  category: Category | null;
  location: Location | null;
  owner: Owner | null;
  macs: Mac[];
  primary_mac: Mac;
}

export interface DeviceRequest {
  name: string | null;
  model: string | null;
  category_id: number | null;
  location_id: number | null;
  owner_id: number | null;
  mac_ids: number[];
}
